### ISSUE TITLE
feat: break templates into parts

### ISSUE THREAD

**Issue by yinonov** [microsoft/fast#5478]
# 🙋 Feature Request

Not sure if this was already discussed / decided upon (didn't find a matching request).

Templates of fast-foundation are composed in a single _template_ method, which makes it hard to override only small portions of it if need be.

to customize the templates library authors need to copy & paste the whole template literal and maintain while making sure nothing breaks over time.

## 🤔 Expected Behavior

breaking down templates to meaningful portions can reduce the complexity and help authors import the needed parts and override some.

so instead of the below code
https://github.com/microsoft/fast/blob/b04dba97fbae2a3e7ed28660c0023984bb29ac2f/packages/web-components/fast-foundation/src/text-field/text-field.template.ts#L24-L79

extraction would look like
```js
html`
  <template...
    ${labelTemplate()}
    ${inputTemplate()}
```

## 💁 Possible Solution

Going the extra mile by dedicating a class for templates would make it more ergonomic, particularly by using typescript's _noImplicitOverride_

something like -
```js
class MyTextfieldTemplate extends FastTextfieldTemplate {
  override labelTemplate() {
    // return...
  }
  // author only needs to override the label html
}
```

**Comment by KingOfTac** [comment]
I was intrigued by this and created a repro on stackblitz for those that want to play around with the idea.

https://stackblitz.com/edit/fast-design-vmccn3?file=index.ts

**Comment by KingOfTac** [comment]
I think a potential issue with this pattern is for it to work well, styles would need to follow a similar approach.

**Comment by yinonov** [comment]
I don't follow, why does it affect styles?

here's a working example
https://github.com/Vonage/vivid-3/blob/c5b9819bbd385515329a73f128d50680ac764b1a/src/components/badge/badge.template.ts#L20-L36

**Comment by KingOfTac** [comment]
I don't think it would break styles if the templating changed to something like this, but I do think the styles could be more efficient if they followed a similar pattern for extending.


Let's say you wanted to extend `TextField` and reuse the styles for it from fast-components as much as possible, however you decide to change the label template from:
```html
<label class="label">
    <slot></slot>
</label>
```
to:
```html
<label class="super-label"
   <slot name="label-icon"></slot>
   <slot></slot>
</label>
```

In this scenario if you want to have custom styles for the label, but keep all of the default styles for the rest of the component, then you would need to do something like this:
```ts
import { textFieldStyles } from "@microsoft/fast-components";

export const myTextFieldStyles = (context, definition) => css`
    ${textFieldStyles}

    .super-label {
        // my custom styles
    }
    
`;
```

This works just fine because you are essentially appending the new styles onto the end of the existing stylesheet. The issue arises when you are extending larger and more complex components, there can become a lot of un-used style declarations in your component which can increase file size and bloat and can also potentially introduce side effects into the component.

In order to get the best of both worlds of extensibility and reusability from fast, then styles should probably mimic the same pattern as the templates, so something like:
```html
<template>
    ${labelTemplate()}
    ${inputTemplate()}
</template>
```

```css
:host { }
${labelStyles()}
${inputStyles()}
```

This way if you intend to swap out the label template for something custom, you can easily do the same for the label's styles, and still reuse the rest of it.

**Comment by yinonov** [comment]
Of course, that's what you would do with sass (mixing etc'). JS has a lot more to offer in that.
But that doesn't mean i'm migrating from sass just yet.

**Comment by yinonov** [comment]
So yeah, you describe a modular system for css as well. I agree. But anyway fast-foundation comes styles-free and the ask is to adjust the foundation templates

**Comment by yinonov** [comment]
can be closed considering #5806

**Comment by chrisdholt** [comment]
> can be closed considering #5806

Closing, thanks @yinonov for this issue and all the feedback - looking forward to making these updates :)
